File: packages/protocol/src/events.ts
// Agent events emitted during execution
export type AgentEventKind =
  | "tool_start"
  | "tool_input"
  | "tool_end"
  | "file_change"
  | "command_start"
  | "command_end"
  | "thinking"
  | "port_detected";

export interface BaseAgentEvent {
  kind: AgentEventKind;
  timestamp: string;
}

export interface ToolStartEvent extends BaseAgentEvent {
  kind: "tool_start";
  tool: string;
  toolUseId: string;
  input: Record<string, unknown>;
}

export interface ToolInputEvent extends BaseAgentEvent {
  kind: "tool_input";
  toolUseId: string;
  inputDelta: string; // Partial JSON being streamed
}

export interface ToolEndEvent extends BaseAgentEvent {
  kind: "tool_end";
  tool: string;
  toolUseId: string;
  result?: string;
  error?: string;
}

export interface FileChangeEvent extends BaseAgentEvent {
  kind: "file_change";
  path: string;
  action: "create" | "edit" | "delete";
  linesAdded?: number;
  linesRemoved?: number;
  diff?: string; // Unified diff content (for future use)
}

export interface CommandStartEvent extends BaseAgentEvent {
  kind: "command_start";
  command: string;
  cwd?: string;
}

export interface CommandEndEvent extends BaseAgentEvent {
  kind: "command_end";
  command: string;
  exitCode: number;
  output?: string;
}

export interface ThinkingEvent extends BaseAgentEvent {
  kind: "thinking";
  content: string;
}

export interface PortDetectedEvent extends BaseAgentEvent {
  kind: "port_detected";
  port: number;
  process?: string;
  previewUrl?: string;
}

export type AgentEvent =
  | ToolStartEvent
  | ToolInputEvent
  | ToolEndEvent
  | FileChangeEvent
  | CommandStartEvent
  | CommandEndEvent
  | ThinkingEvent
  | PortDetectedEvent;

File: apps/control-plane/internal/protocol/events.go
package protocol

type AgentEventKind string

const (
	EventKindToolStart    AgentEventKind = "tool_start"
	EventKindToolInput    AgentEventKind = "tool_input"
	EventKindToolEnd      AgentEventKind = "tool_end"
	EventKindFileChange   AgentEventKind = "file_change"
	EventKindCommandStart AgentEventKind = "command_start"
	EventKindCommandEnd   AgentEventKind = "command_end"
	EventKindThinking     AgentEventKind = "thinking"
	EventKindPortDetected AgentEventKind = "port_detected"
)

// AgentEvent is a polymorphic event type. Fields are optional based on Kind.
type AgentEvent struct {
	Kind       AgentEventKind         `json:"kind"`
	Timestamp  string                 `json:"timestamp"`
	Tool       string                 `json:"tool,omitempty"`
	ToolUseID  string                 `json:"toolUseId,omitempty"`
	Input      map[string]interface{} `json:"input,omitempty"`
	InputDelta string                 `json:"inputDelta,omitempty"`
	Result     *string                `json:"result,omitempty"`
	Error      *string                `json:"error,omitempty"`
	Path      string                 `json:"path,omitempty"`
	Action    string                 `json:"action,omitempty"`
	Command   string                 `json:"command,omitempty"`
	Cwd       *string                `json:"cwd,omitempty"`
	ExitCode  *int                   `json:"exitCode,omitempty"`
	Output    *string                `json:"output,omitempty"`
	Content   string                 `json:"content,omitempty"`
	Port      int                    `json:"port,omitempty"`
	Process   *string                `json:"process,omitempty"`
	PreviewURL *string               `json:"previewUrl,omitempty"`
	LinesAdded   *int                `json:"linesAdded,omitempty"`
	LinesRemoved *int                `json:"linesRemoved,omitempty"`
}
